storage: panic in bestRebalanceTarget

My cluster crashed during a partitioning of TPCC-10k. Assigning to @a-robinson based on `git blame` info.

Build tag: `v2.0-beta.20180305-231-gc422be5`

```
panic(0x1fa1160, 0x3322ea0)
        /usr/lib/go-1.10/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/storage.bestRebalanceTarget(0xc42029f070, 0xc420a887e0, 0xc433eed680, 0x2, 0x3, 0x4037000000000000, 0x40bbb42c8590b216, 0x4037000000000000, 0x40a2546f4de9bd38)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/allocator_scorer.go:803 +0x269
github.com/cockroachdb/cockroach/pkg/storage.Allocator.RebalanceTarget(0xc4202b4780, 0xc4202cc9d0, 0xc42029f070, 0xc420a887e0, 0x250d4a0, 0xc426118660, 0x100000, 0x4000000, 0x300015f90, 0xc4506ec820, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/allocator.go:530 +0x54c
github.com/cockroachdb/cockroach/pkg/storage.(*replicateQueue).shouldQueue(0xc42002a230, 0x250d4a0, 0xc426118660, 0x151b8c878a1acfae, 0x900000000, 0xc427669500, 0xc42b730000, 0x58, 0x80, 0x17, ...)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/replicate_queue.go:183 +0x67e
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).maybeAddLocked(0xc42023a800, 0x250d4a0, 0xc426118660, 0xc427669500, 0x151b8c878a1acfae, 0xc400000000)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:428 +0x144
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).MaybeAdd(0xc42023a800, 0xc427669500, 0x151b8c878a1acfae, 0x0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:381 +0xd9
github.com/cockroachdb/cockroach/pkg/storage.(*replicaScanner).waitAndProcess(0xc4206a40d0, 0x250d4a0, 0xc426c84ae0, 0xc4205a8360, 0xde52ba3, 0xed23a075b, 0x0, 0xc427669500, 0x1d00)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/scanner.go:226 +0x30b
github.com/cockroachdb/cockroach/pkg/storage.(*replicaScanner).scanLoop.func1.1(0xc427669500, 0xc420a88501)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/scanner.go:274 +0x81
github.com/cockroachdb/cockroach/pkg/storage.(*storeReplicaVisitor).Visit(0xc420a885a0, 0xc441b21940)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:324 +0x1a6
```

Yeah, it was an NPE - sorry for not including that. The log has a lot of remote declined snapshots. I didn't think that was related, but just in case I'll paste a bunch of context.

```
I180313 17:56:46.140538 1529619 storage/replica_raftstorage.go:520  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] generated preemptive snapshot 4b4acfd9 at index 92
I180313 17:56:46.140571 1529619 storage/replicate_queue.go:226  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] snapshot failed: breaker open
I180313 17:56:46.313548 1529619 storage/replica_raftstorage.go:520  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] generated preemptive snapshot 8f9d5006 at index 92
I180313 17:56:46.313592 1529619 storage/replicate_queue.go:226  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] snapshot failed: breaker open
I180313 17:56:46.732969 1529619 storage/replica_raftstorage.go:520  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] generated preemptive snapshot b50d9629 at index 92
I180313 17:56:46.733007 1529619 storage/replicate_queue.go:226  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] snapshot failed: breaker open
W180313 17:56:46.760101 1529528 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0023:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180313 17:56:46.760157 1529528 vendor/google.golang.org/grpc/clientconn.go:830  Failed to dial jordan-tpcc-0023:26257: context canceled; please retry.
I180313 17:56:47.446277 1529619 storage/replica_raftstorage.go:520  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] generated preemptive snapshot 792b1801 at index 92
W180313 17:56:47.447429 1529614 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0023:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 10.142.0.12:26257: connect: connection refused". Reconnecting...
I180313 17:56:47.447539 1529619 storage/replicate_queue.go:226  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] snapshot failed: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
E180313 17:56:47.447558 1529619 storage/queue.go:778  [replicate,n12,s9,r27801/2:/Table/55/1/793{4/5/-…-6/4/-…}] failed to replicate after 5 retries
I180313 17:56:47.449522 1529629 storage/replica_raftstorage.go:520  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] generated preemptive snapshot c448948a at index 67
W180313 17:56:48.386455 1765 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0006:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180313 17:56:48.386460 1480276 storage/raft_transport.go:465  [n12] raft transport stream to node 21 failed: rpc error: code = Unavailable desc = transport is closing
W180313 17:56:48.386486 1765 vendor/google.golang.org/grpc/clientconn.go:1277  grpc: addrConn.transportMonitor exits due to: context canceled
W180313 17:56:48.446648 1529614 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0023:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180313 17:56:48.446687 1529614 vendor/google.golang.org/grpc/clientconn.go:830  Failed to dial jordan-tpcc-0023:26257: context canceled; please retry.
I180313 17:56:51.080514 1529629 storage/store.go:3711  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] streamed snapshot to (n13,s13):?: kv pairs: 106561, log entries: 29, rate-limit: 2.0 MiB/sec, 3631ms
I180313 17:56:51.152278 1529629 storage/replica_command.go:1760  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] change replicas (ADD_REPLICA (n13,s13):4): read existing descriptor r20959:/Table/55/1/548{8/9/-1823/1/0-9/2/-179/8} [(n4,s4):1, (n12,s9):2, (n24,s24):3, next=4]
I180313 17:56:51.173425 1529877 storage/replica.go:3209  [n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] proposing ADD_REPLICA((n13,s13):4): updated=[(n4,s4):1 (n12,s9):2 (n24,s24):3 (n13,s13):4] next=5
I180313 17:56:51.178405 1529867 storage/replica_command.go:1760  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] change replicas (REMOVE_REPLICA (n24,s24):3): read existing descriptor r20959:/Table/55/1/548{8/9/-1823/1/0-9/2/-179/8} [(n4,s4):1, (n12,s9):2, (n24,s24):3, (n13,s13):4, next=5]
I180313 17:56:51.200386 1529892 storage/replica.go:3209  [n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] proposing REMOVE_REPLICA((n24,s24):3): updated=[(n4,s4):1 (n12,s9):2 (n13,s13):4] next=5
I180313 17:56:51.205102 1529894 storage/replica_raftstorage.go:520  [replicate,n12,s9,r16612/3:/Table/56/1/144{0/823…-1/865…}] generated preemptive snapshot be55b65c at index 147
I180313 17:56:51.205823 1529894 storage/replicate_queue.go:226  [replicate,n12,s9,r16612/3:/Table/56/1/144{0/823…-1/865…}] snapshot failed: (n17,s16):?: remote declined snapshot: store busy applying snapshots and/or removing replicas
I180313 17:56:51.260466 1529878 storage/replica_raftstorage.go:520  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] generated preemptive snapshot b4bed331 at index 73
I180313 17:56:51.261092 1529878 storage/replicate_queue.go:226  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] snapshot failed: (n15,s15):?: remote declined snapshot: store busy applying snapshots and/or removing replicas
I180313 17:56:51.317053 1529878 storage/replica_raftstorage.go:520  [replicate,n12,s9,r20959/2:/Table/55/1/548{8/9/-…-9/2/-…}] generated preemptive snapshot e04281a5 at index 73
W180313 17:56:51.681134 1516492 storage/raft_transport.go:465  [n12] raft transport stream to node 14 failed: rpc error: code = Unavailable desc = transport is closing
W180313 17:56:51.681227 31 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0007:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180313 17:56:51.681282 31 vendor/google.golang.org/grpc/clientconn.go:1277  grpc: addrConn.transportMonitor exits due to: context canceled
W180313 17:56:52.479726 2845 vendor/google.golang.org/grpc/clientconn.go:1158  grpc: addrConn.createTransport failed to connect to {jordan-tpcc-0016:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180313 17:56:52.479731 2841 storage/raft_transport.go:465  [n12] raft transport stream to node 11 failed: rpc error: code = Unavailable desc = transport is closing
W180313 17:56:52.479771 2845 vendor/google.golang.org/grpc/clientconn.go:1277  grpc: addrConn.transportMonitor exits due to: context canceled
E180313 17:56:52.933403 568 util/log/crash_reporting.go:182  [n12] a panic has occurred!
E180313 17:56:52.942814 568 util/log/crash_reporting.go:440  [n12] Reported as error 72acc16970cf45c195ea1d974c986cce
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x138cc39]
```

--- a/pkg/storage/allocator_scorer.go
+++ b/pkg/storage/allocator_scorer.go
@@ -787,6 +787,9 @@ func bestRebalanceTarget(
 			continue
 		}
 		target := option.candidates.selectGood(randGen)
+		if target == nil {
+			continue
+		}
 		existing := option.existingCandidates[len(option.existingCandidates)-1]
 		if betterRebalanceTarget(target, &existing, bestTarget, &replaces) == target {
 			bestIdx = i

--- a/pkg/storage/allocator_scorer_test.go
+++ b/pkg/storage/allocator_scorer_test.go
@@ -307,6 +307,100 @@ func TestBetterThan(t *testing.T) {
 	}
 }
 
+// TestBestRebalanceTarget constructs a hypothetical output of
+// rebalanceCandidates and verifies that bestRebalanceTarget properly returns
+// the candidates in the ideal order of preference and omits any that aren't
+// desirable.
+func TestBestRebalanceTarget(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	candidates := []rebalanceOptions{
+		{
+			candidates: []candidate{
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 11},
+					valid:          true,
+					necessary:      true,
+					diversityScore: 1.0,
+					rangeCount:     11,
+				},
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 12},
+					valid:          true,
+					necessary:      true,
+					diversityScore: 1.0,
+					rangeCount:     12,
+				},
+			},
+			existingCandidates: []candidate{
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 1},
+					valid:          true,
+					necessary:      true,
+					diversityScore: 0,
+					rangeCount:     1,
+				},
+			},
+		},
+		{
+			candidates: []candidate{
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 13},
+					valid:          true,
+					necessary:      true,
+					diversityScore: 1.0,
+					rangeCount:     13,
+				},
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 14},
+					valid:          false,
+					necessary:      false,
+					diversityScore: 0.0,
+					rangeCount:     14,
+				},
+			},
+			existingCandidates: []candidate{
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 2},
+					valid:          true,
+					necessary:      false,
+					diversityScore: 0,
+					rangeCount:     2,
+				},
+				{
+					store:          roachpb.StoreDescriptor{StoreID: 3},
+					valid:          false,
+					necessary:      false,
+					diversityScore: 0,
+					rangeCount:     3,
+				},
+			},
+		},
+	}
+
+	expected := []roachpb.StoreID{13, 11, 12}
+	allocRand := makeAllocatorRand(rand.NewSource(0))
+	var i int
+	for {
+		i++
+		target, existing := bestRebalanceTarget(allocRand, candidates)
+		if len(expected) == 0 {
+			if target == nil {
+				break
+			}
+			t.Errorf("round %d: expected nil, got target=%+v, existing=%+v", i, target, existing)
+			continue
+		}
+		if target == nil {
+			t.Errorf("round %d: expected s%d, got nil", i, expected[0])
+		} else if target.store.StoreID != expected[0] {
+			t.Errorf("round %d: expected s%d, got target=%+v, existing=%+v",
+				i, expected[0], target, existing)
+		}
+		expected = expected[1:]
+	}
+}
+
 func TestStoreHasReplica(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

